rust/worker/src/index/fulltext/types.rs

```diff
@@ -116,227 +50,605 @@ impl FullTextIndex for BlockfileFullTextIndex {
+            .posting_lists_blockfile_reader
+            .get_by_prefix(first_token)
+            .unwrap();
+        for (_, doc_id, positions) in first_token_positional_posting_list.iter() {
             let positions_vec: Vec<i32> = positions
-                .values()
                 .iter()
-                .map(|x| *x - first_token_offset)
+                .map(|x| x.unwrap() - first_token_offset)
                 .collect();
             candidates.insert(*doc_id, positions_vec);
         }
 
         // Iterate through the rest of the tokens, intersecting the posting lists with the candidates.
+        let mut token_offset = 0;
         for (token, _) in token_frequencies[1..].iter() {
-            let blockfilekey = BlockfileKey::new("".to_string(), Key::String(token.to_string()));
-            let positional_posting_list =
-                match self.posting_lists_blockfile.get(blockfilekey).unwrap() {
-                    Value::PositionalPostingListValue(arr) => arr,
-                    _ => panic!("Value is not an arrow struct array"),
-                };
-            let token_offset = tokens
-                .iter()
-                .find(|t| t.text == *token)
-                .unwrap()
-                .offset_from as i32;
-            let mut new_candidates: HashMap<i32, Vec<i32>> = HashMap::new();
+            token_offset += 1;
+            let positional_posting_list = self
+                .posting_lists_blockfile_reader
+                .get_by_prefix(token.as_str())
+                .unwrap();
+            // TODO once we sort by frequency, we need to find the token position
+            // here, taking into account which positions for repeats of the same
+            // token have already been used up.
+            // let token_offset = tokens
+            //     .iter()
+            //     .find(|t| t.text == *token)
+            //     .unwrap()
+            //     .offset_from as i32;
+            let mut new_candidates: HashMap<u32, Vec<i32>> = HashMap::new();
             for (doc_id, positions) in candidates.iter() {
                 let mut new_positions = vec![];
                 for position in positions {
-                    if let Some(positions_for_doc_id) =
-                        positional_posting_list.get_positions_for_doc_id(*doc_id)
+                    if let Some(positions) =
+                        // Find the positional posting list with second item = to doc_id
+                        positional_posting_list
+                            .iter()
+                            .find(|x| x.1 == *doc_id)
+                            .map(|x| &x.2)
                     {
-                        for position_for_doc_id in positions_for_doc_id.values() {
-                            if position_for_doc_id - token_offset == *position {
+                        for pos in positions.iter() {
+                            if pos.unwrap() == position + token_offset {
```

Comment:
is this unwrap safe?

Reply:
If it's not it's an invariant violation. Changed to a `Result`